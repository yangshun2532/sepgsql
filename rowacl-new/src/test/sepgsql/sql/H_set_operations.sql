--
-- SE-PostgreSQL testcases : Set Operations/WITH RECURSIVE
--
-- selinux_user: unconfined_u
-- selinux_role: unconfined_r
-- selinux_type: sepgsql_test_t
-- selinux_range: s0:c0

WITH RECURSIVE h9 AS
(
	SELECT security_context, * FROM h1 WHERE pid is NULL

	UNION ALL

	SELECT h1.security_context, h1.* FROM h1,h9 WHERE h1.pid = h9.id
)
SELECT * FROM h9;

WITH h9 AS (SELECT * FROM h2)
SELECT * FROM h9;		-- to be failed

WITH h9 AS (SELECT t FROM h2)
SELECT t FROM h9;

SELECT * FROM h3
UNION
SELECT * FROM h4;

SELECT * FROM h3
INTERSECT
SELECT * FROM h4;

SELECT * FROM h3
EXCEPT
SELECT * FROM h4;


