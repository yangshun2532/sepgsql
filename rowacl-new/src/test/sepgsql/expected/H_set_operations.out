--
-- SE-PostgreSQL testcases : Set Operations/WITH RECURSIVE
--
-- selinux_user: unconfined_u
-- selinux_role: unconfined_r
-- selinux_type: sepgsql_test_t
-- selinux_range: s0:c0
WITH RECURSIVE h9 AS
(
	SELECT security_context, * FROM h1 WHERE pid is NULL
	UNION ALL
	SELECT h1.security_context, h1.* FROM h1,h9 WHERE h1.pid = h9.id
)
SELECT * FROM h9;
           security_context           | id  | pid |  name  
--------------------------------------+-----+-----+--------
 system_u:object_r:sepgsql_table_t:s0 |   0 |     | /
 system_u:object_r:sepgsql_table_t:s0 |   1 |   0 | /a
 system_u:object_r:sepgsql_table_t:s0 |  11 |   1 | /a/a
 system_u:object_r:sepgsql_table_t:s0 | 111 |  11 | /a/a/a
 system_u:object_r:sepgsql_table_t:s0 | 113 |  11 | /a/a/c
(5 rows)

WITH h9 AS (SELECT * FROM h2)
SELECT * FROM h9;		-- to be failed
psql:./H_set_operations.sql:20: ERROR:  SELinux: denied { select } scontext=unconfined_u:unconfined_r:sepgsql_test_t:s0:c0 tcontext=system_u:object_r:sepgsql_secret_table_t:s0 tclass=db_column name=h2.s
WITH h9 AS (SELECT t FROM h2)
SELECT t FROM h9;
 t 
---
(0 rows)

SELECT * FROM h3
UNION
SELECT * FROM h4;
 a |  b  
---+-----
 2 | bbb
 1 | aaa
 3 | ccc
 5 | eee
 6 | fff
(5 rows)

SELECT * FROM h3
INTERSECT
SELECT * FROM h4;
 a |  b  
---+-----
 5 | eee
 2 | bbb
(2 rows)

SELECT * FROM h3
EXCEPT
SELECT * FROM h4;
 a |  b  
---+-----
 1 | aaa
(1 row)

